x/lockup/keeper/lock.go

```diff
@@ -536,3 +537,40 @@ func (k Keeper) unlockInternalLogic(ctx sdk.Context, lock types.PeriodLock) erro
 	k.hooks.OnTokenUnlocked(ctx, owner, lock.ID, lock.Coins, lock.Duration, lock.EndTime)
 	return nil
 }
+
+func (k Keeper) EditLockup(ctx sdk.Context, lock types.PeriodLock, newDuration time.Duration) error {
+	// sanity check
+
+	// check unlocking
+	if lock.IsUnlocking() {
+		return fmt.Errorf("cannot edit unlocking lockup")
+	}
+
+	// check synthetic lockup exists
+	if k.HasAnySyntheticLockups(ctx, lock.ID) {
+		return fmt.Errorf("cannot edit lockup with synthetic lock")
+	}
+
+	if newDuration != 0 {
+		// check newDuration > duration
+		if !(newDuration > lock.Duration) {
+			return fmt.Errorf("new duration should be greater than the original")
+		}
+
+		// update accumulation store
+		for _, coin := range lock.Coins {
+			k.accumulationStore(ctx, coin.Denom).Decrease(accumulationKey(lock.Duration), coin.Amount)
+			k.accumulationStore(ctx, coin.Denom).Increase(accumulationKey(newDuration), coin.Amount)
+		}
+
+		lock.Duration = newDuration
+	}
+
+	// update lockup
+	err := k.setLockAndResetLockRefs(ctx, lock)
```

Comment:
Do we not need to run `deleteLockRefs` to delete old lock refs? I'm not quite sure if `setLockAndResetLockRefs` is actually deleting the old refs

Reply:
Thanks for catching it! Will fix this

Reply:
Matt pointed out that `setLockAndResetLockRefs` is bit misleading, because the name and the comment
```
// setLockAndResetLockRefs sets the lock, and resets all of its lock references
// This puts the lock into a 'clean' state, aside from the AccumulationStore.
```
implies that it removes the existing lockrefs, but actually it doesn't. Should we fix this behaviour, or change the name of the method?

Reply:
I think we wouldn't need to fix this within this PR, but instead create an issue to investigate on this.

That being said, I think it would suffice for this PR to ensure that old lock refs are getting deleted using `deleteLockRefs`

Reply:
I have replaced setLockAnd.. to a pair of delete and add lockrefs. This works!

Reply:
Oh interesting, lets make an issue and investigate this

Reply:
https://github.com/osmosis-labs/osmosis/issues/1275